### PR TITLE
fix(v4-sdk): V4 planner accommodates exactOut + granular take/settles

### PR DESCRIPTION
**Pull request by ewilz**

### CODE REVIEW

**Review by marktoda — APPROVED** [review]
quick approval to unblock since requires 2

File: sdks/v4-sdk/src/utils/v4Planner.test.ts [hunk 1]
```diff
@@ -48,69 +48,186 @@ const DAI_USDC = new Pool(
...
+  0,
+  TICKLIST
+)
 
 describe('RouterPlanner', () => {
   let planner: V4Planner
 
   beforeEach(() => {
     planner = new V4Planner()
+    // console.log(planner)
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/v4Planner.test.ts:68]: do you need this still?

File: sdks/v4-sdk/src/utils/v4Planner.ts [hunk 2]
```diff
@@ -91,16 +93,20 @@ const ABI_DEFINITION: { [key in Actions]: string[] } = {
...
   [Actions.TAKE_PORTION]: ['address', 'address', 'uint256'],
   [Actions.SETTLE_TAKE_PAIR]: ['address', 'address'],
   [Actions.CLOSE_CURRENCY]: ['address'],
   [Actions.SWEEP]: ['address', 'address'],
 }
 
+const FULL_DELTA_AMOUNT = 0
+const MSG_SENDER = '0x0000000000000000000000000000000000000001'
+const ADDRESS_THIS = '0x0000000000000000000000000000000000000002'
+
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/v4Planner.ts:109]: ooh nice I made an ActionsConstant utils file, I can put these in there later when my PR gets merged

File: sdks/v4-sdk/src/utils/v4Planner.ts [hunk 3]
```diff
@@ -117,23 +123,41 @@ export class V4Planner {
...
-      ])
-    }
-
-    this.addAction(Actions.SETTLE_TAKE_PAIR, [currencyIn, currencyOut])
+    this.addAction(actionType, [
+      exactOutput
+        ? {
+            currencyOut,
+            path: encodeRouteToPath(trade.route, exactOutput),
+            amountInMaximum: trade.maximumAmountIn(slippageTolerance ?? new Percent(0)).quotient.toString(),
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/v4Planner.ts:141]: it looks like on 128 we check that slippageTolerance is required? so can we just pass it in directly here?

File: sdks/v4-sdk/src/utils/v4Planner.ts [hunk 4]
```diff
@@ -117,23 +123,41 @@ export class V4Planner {
   }
 
   addTrade(trade: Trade<Currency, Currency, TradeType>, slippageTolerance?: Percent): void {
-    const actionType = trade.tradeType === TradeType.EXACT_INPUT ? Actions.SWAP_EXACT_IN : Actions.SWAP_EXACT_OUT
+    const exactOutput = trade.tradeType === TradeType.EXACT_OUTPUT
+
+    if (exactOutput) invariant(!!slippageTolerance, 'ExactOut requires slippageTolerance')
```
Comment (snreynolds) [inline sdks/v4-sdk/src/utils/v4Planner.ts:128]: how come we only check this for exactOuput? just curious

File: sdks/v4-sdk/src/utils/v4Planner.ts (hunk 2 above)
Comment (ewilz) [reply sdks/v4-sdk/src/utils/v4Planner.ts:109]: yah feel free to move around. this file is def pretty cray

File: sdks/v4-sdk/src/utils/v4Planner.ts (hunk 3 above)
Comment (ewilz) [reply sdks/v4-sdk/src/utils/v4Planner.ts:141]: typescript gets mad

File: sdks/v4-sdk/src/utils/v4Planner.ts (hunk 4 above)
Comment (ewilz) [reply sdks/v4-sdk/src/utils/v4Planner.ts:128]: good question I can leave a comment. on exactInput sometimes we do agg slippage checks so we don't need to know slippage, but we enver do that on exactOutput

File: sdks/v4-sdk/src/utils/v4Planner.ts [hunk 5]
```diff
@@ -117,23 +123,42 @@ export class V4Planner {
...
+        : {
+            currencyIn,
+            path: encodeRouteToPath(trade.route, exactOutput),
+            amountIn: trade.inputAmount.quotient.toString(),
+            amountOutMinimum: slippageTolerance ? trade.minimumAmountOut(slippageTolerance).quotient.toString() : 0,
+          },
+    ])
+  }
+
+  addSettle(currency: Currency, payerIsUser: boolean, amount?: BigNumber): void {
```
Comment (codyborn) [inline sdks/v4-sdk/src/utils/v4Planner.ts:154]: nit: should we add comment headers?

**Review by codyborn — APPROVED** [review]
LGTM!

File: sdks/v4-sdk/src/utils/v4Planner.ts (hunk 3 above)
Comment (snreynolds) [reply sdks/v4-sdk/src/utils/v4Planner.ts:142]: hahah ok got it

**Review by snreynolds — APPROVED** [review]
looks good

**Pull request merged**